# Some Tips

## fmt.Printf()、数组的数据类型
输出一个值的类型

以下为数组在 Go 和 C 中的主要区别。在 Go 中,
- 数组是值。 将一个数组赋值给另一个数组会复制其所有元素。 （值传递，而非引用传递）
- 特别地,若将某个数组传入某个函数,它将接收到该数组的一份副本而非指针。 
- 数组的大小是其类型的一部分。类型 [10]int 和 [20]int 是不同的。

```go
package main

import (
	"fmt"
)

func main() {
	s := [3]int{1, 2, 3}
	fmt.Printf("%T ", s)
}
```
结果是
```
[3]int 
```

## defer语句
defer 语句用于预设一个函数调用(即推迟执行函数)。

被推迟函数的`实参`(如果该函数为方法则还包括接收者)在`声明时`就会求值, 而不是在真正调用执行时才求值。 
这样无需担心变量值在函数执行时被改变。 但被推迟函数的函数体是在最后执行的时候才会执行。

在下面的例子中，声明` defer un(trace("a"))`时，un()函数是作为被推迟函数，那么un()函数的参数在声明的时候就会进行求值，而不是到真正执行un()函数时再去求值。

```go
func trace(s string) string {
    fmt.Println("entering:", s)
    return s
}
func un(s string) {
    fmt.Println("leaving:", s)
}
func a() {
    defer un(trace("a"))
    fmt.Println("in a")
}
func b() {
    defer un(trace("b"))
    fmt.Println("in b")
    a()
}
func main() {
    b()
}
```
输出如下
```
entering: b
in b
entering: a
in a
leaving: a
leaving: b
```

再看一个例子，其输出是 132
```go
package main

import (
	"fmt"
)

type Slice []int

func NewSlice() Slice {
	return make(Slice, 0)
}
func (s *Slice) Add(elem int) *Slice {
	*s = append(*s, elem)
	fmt.Print(elem)
	return s
}
func main() {
	s := NewSlice()
	defer s.Add(1).Add(2)
	s.Add(3)
}
```

defer语句有可能改变函数的返回值，下面例子将返回 2
```go
func c() (i int) {
    defer func() { i++ }()
    return 1
}
```

## golang中的nil
1. nil没有type
2. 在Go语言中，未初始化的变量拥有其类型的zero value。 
3. pointer，slice，map，channel，function和interface 6种类型的零值是 nil

```
类型	                     nil值含义
pointer	                 指向nothing
slice	                 slice变量中的3个成员值：buf为nil（没有backing array），len和cap都是0
map，channel，function	 一个nil pointer，指向nothing
interface	             interface包含”type, value”，一个nil interface必须二者都为nil: ”nil, nil”
```

nil只能赋值给指针、channel、func、interface、map或slice类型的变量。
如果将nil赋值给其他类型变量的时候将会引发panic。
```go
package main

import (
	"fmt"
	"reflect"
)

func main() {
	fmt.Printf("Func type nil:%#v\n", (func())(nil))
	fmt.Printf("Map type nil:%#v\n", map[string]string(nil))
	fmt.Printf("Slice type nil:%#v\n", []string(nil))
	fmt.Printf("Interface{} type nil:%#v\n", nil)
	fmt.Printf("Channel type nil:%#v\n", (chan struct{})(nil))
	fmt.Printf("Pointer type nil:%#v\n", (*struct{})(nil))
	fmt.Printf("Pointer type nil:%#v\n", (*int)(nil))

	//将一个nil对象赋值给一个interface，这个interface为非nil。这是因为这个对象的类型不是nil
	fmt.Println(nil == interface{}((func())(nil))) //false
	fmt.Println(reflect.TypeOf(interface{}((func())(nil))))
}
```
输出如下
```
Func type nil:(func())(nil)
Map type nil:map[string]string(nil)
Slice type nil:[]string(nil)
Interface{} type nil:<nil>
Channel type nil:(chan struct {})(nil)
Pointer type nil:(*struct {})(nil)
Pointer type nil:(*int)(nil)
false
func()
```

见[理解Go中的nil](https://studygolang.com/topics/2863)，[understanding-golang-nil-value](https://www.gmarik.info/blog/2016/understanding-golang-nil-value/)

## panic
当 panic 被调用后(包括不明确的运行时错误,例如切片检索越界或类型断言失败), 
程序将立刻终止当前函数的执行,并开始回溯Go程的栈, 运行任何被推迟的函数。
若回溯到达 Go 程栈的顶端,程序就会终止。

如果在一个groutine中发生了panic， 定义在panic后的语句都不会被执行到，包括defer语句。 
只有定义在panic语句前面的defer语句才会在回溯过程中执行。

所以recover()语句必须放在panic()前面的defer语句中，才会生效。

recover()语句负责处理panic产生的错误。

关键的一点是，即使函数执行的时候panic了，函数不往下走了，运行时并不是立刻向上传递panic，而是到defer那，等defer的东西都跑完了，panic再向上传递。所以这时候 defer 有点类似 try-catch-finally 中的 finally。 

换句话说，一旦panic，逻辑就会走到defer那，那我们就在defer那等着，调用recover函数将会捕获到当前的panic（如果有的话），被捕获到的panic就不会向上传递了，于是，世界恢复了和平。你可以干你想干的事情了。

不过要注意的是，recover之后，逻辑并不会恢复到panic那个点去，函数还是会在defer之后返回。

Go中可以抛出一个panic的异常，然后在defer中通过recover捕获这个异常，然后正常处理。

```go
package main
package main

import "fmt"

func main() {
	f()
	fmt.Println("Returned normally from f.")
}

func f() {
	defer func() {
		if r := recover(); r != nil {
			fmt.Println("Recovered in f", r)
		}
	}()
	fmt.Println("Calling g.")
	g(0)
	fmt.Println("Returned normally from g.")
}

func g(i int) {
	if i > 1 {
		fmt.Println("Panicking!")
		panic(fmt.Sprintf("%v", i))
	}
	defer fmt.Println("Defer in g", i)
	fmt.Println("Printing in g", i)
	g(i + 1)
}
```

输出如下，
```
Calling g.
Printing in g 0
Printing in g 1
Panicking!
Defer in g 1
Defer in g 0
Recovered in f 2
Returned normally from f.
```

如果把上面代码中的`recover()`删除了，输出如下：
```
Calling g.
Printing in g 0
Printing in g 1
Panicking!
Defer in g 1
Defer in g 0
panic: 2

goroutine 1 [running]:
main.g(0x2)
	/Users/fanqihong/Desktop/go-project/src/ftmtest/fmttest.go:19 +0x3b6
main.g(0x1)
	/Users/fanqihong/Desktop/go-project/src/ftmtest/fmttest.go:23 +0x223
main.g(0x0)
	/Users/fanqihong/Desktop/go-project/src/ftmtest/fmttest.go:23 +0x223
main.f()
	/Users/fanqihong/Desktop/go-project/src/ftmtest/fmttest.go:12 +0xa3
main.main()
	/Users/fanqihong/Desktop/go-project/src/ftmtest/fmttest.go:6 +0x26
```

## 追加方式写文件
```go
func write_log(st string){
        f,err:=os.OpenFile("/root/log.txt",os.O_WRONLY|os.O_APPEND, 0666)
        defer f.Close()
        if err!=nil {
                return
        }
        f.WriteString(st+"\n")
}
```

## Reflect用法
根据hanshumin
```
package main

import (
	"flag"
	"fmt"
	"reflect"
)

func main() {
	flag.Parse()
	a := MyStruct{}
	comamd, err := a.run(flag.Args()...)
	if err == nil {
		comamd()
	}

}

type MyStruct struct {
}

func (a MyStruct) run(args ...string) (func(...string) error, error) {
	if len(args) > 0 {
		methodname := "Cmd" + args[0]
		fmt.Println(methodname)
		//		method := reflect.ValueOf(a).MethodByName("Cmdps")
		method := reflect.ValueOf(a).MethodByName(methodname)
		fmt.Println(method)
		if method.IsValid() {
			fmt.Println("yes")
			return method.Interface().(func(...string) error), nil
		} else {
			fmt.Println("no")
			return nil, nil
		}
	}
	return nil, nil
}

func (a MyStruct) Cmdversion(args ...string) error {
	fmt.Println("cmd version methond")
	return nil
}

func (a MyStruct) Cmdps(args ...string) error {
	fmt.Println("cmd ps methond")
	return nil
}
```

## go groutine的理解
第一个循环不会完全输出0，1，2。 因为外面的i会很快就会加到3，但groutine还没有全部起来。
```go
package main

import (
	"fmt"
	"runtime"
	"sync"
)

func main() {
	runtime.GOMAXPROCS(1)
	wg := sync.WaitGroup{}
	wg.Add(6)
	for i := 0; i < 3; i++ {
		go func() {
			fmt.Println("i: ", i)
			wg.Done()
		}()
	}
	for i := 0; i < 3; i++ {
		go func(i int) {
			fmt.Println("i: ", i)
			wg.Done()
		}(i)
	}
	wg.Wait()
}
```
